crates/ra_hir/src/code_model.rs

```diff
@@ -960,6 +960,38 @@ impl ImplDef {
         db.impl_data(self.id).target_trait.clone()
     }
 
+    pub fn target_trait_substs_matches(&self, db: &dyn HirDatabase, typs: &[Type]) -> bool {
+        let type_ref = match self.target_trait(db) {
+            Some(typ_ref) => typ_ref,
+            None => return false,
+        };
+        let resolver = self.id.resolver(db.upcast());
+        let ctx = hir_ty::TyLoweringContext::new(db, &resolver);
+        let ty = Ty::from_hir(&ctx, &type_ref);
+        let d = match ty.dyn_trait_ref() {
+            Some(d) => d,
+            None => return false,
+        };
+        let mut matches = true;
+        let mut i = 0;
+        d.substs.walk(&mut |t| {
+            if matches {
+                if i >= typs.len() {
+                    matches = false;
+                    return;
+                }
+                match t {
+                    Ty::Bound(_) => matches = i == 0,
```

Comment:
The first Bound here is the enum, I believe? I'm not sure I completely understood this bit of the API so I think this method suffers from that!